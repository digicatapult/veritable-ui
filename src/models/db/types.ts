import { Knex } from 'knex'
import { z } from 'zod'
import { carbonEmbodimentRequestData, carbonEmbodimentResponseData } from '../drpc.js'

export const tablesList = [
  'connection',
  'connection_invite',
  'query',
  'query_rpc',
  'settings',
  'organisation_registries',
] as const

const insertConnection = z.object({
  company_name: z.string(),
  company_number: z.string(),
  status: z.union([
    z.literal('pending'),
    z.literal('unverified'),
    z.literal('verified_them'),
    z.literal('verified_us'),
    z.literal('verified_both'),
    z.literal('disconnected'),
  ]),
  agent_connection_id: z.union([z.string(), z.null()]),
  pin_attempt_count: z.number().int().gte(0).lte(255),
  pin_tries_remaining_count: z.number().int().gte(0).lte(255).nullable(),
  registry_country_code: z.string(),
})

const insertOrganisationRegistries = z.object({
  country_code: z.string(),
  registry_name: z.string(),
  registry_key: z.string(),
  url: z.string(),
  api_key: z.string(),
})
const insertConnectionInvite = z.object({
  connection_id: z.string(),
  oob_invite_id: z.string(),
  pin_hash: z.string(),
  expires_at: z.date(),
  validity: z.union([z.literal('valid'), z.literal('expired'), z.literal('too_many_attempts'), z.literal('used')]),
})

const insertQueryShared = z.object({
  connection_id: z.string(),
  parent_id: z.string().nullable().optional(),
  type: z.literal('total_carbon_embodiment'),
  status: z.enum(['resolved', 'pending_your_input', 'pending_their_input', 'errored', 'forwarded']),
  response_id: z.string().nullable(),
  role: z.enum(['requester', 'responder']),
  expires_at: z.date(),
})
const totalCarbonEmbodimentQuery = z.object({
  type: z.literal('total_carbon_embodiment'),
  details: carbonEmbodimentRequestData,
  response: carbonEmbodimentResponseData.nullable(),
})
const insertQuery = z.discriminatedUnion('type', [totalCarbonEmbodimentQuery.merge(insertQueryShared)])

const defaultFields = z.object({
  id: z.string(),
  created_at: z.date(),
  updated_at: z.date(),
})

const insertQueryRpc = z.object({
  query_id: z.string(),
  agent_rpc_id: z.string(),
  role: z.union([z.literal('client'), z.literal('server')]),
  method: z.union([z.literal('submit_query_request'), z.literal('submit_query_response')]),
  result: z.union([z.record(z.any()), z.null()]).optional(),
  error: z.union([z.record(z.any()), z.null()]).optional(),
})
const insertSettings = z.object({
  setting_key: z.string(),
  setting_value: z.string(),
})

const Zod = {
  connection: {
    insert: insertConnection,
    get: insertConnection.merge(defaultFields),
  },
  connection_invite: {
    insert: insertConnectionInvite,
    get: insertConnectionInvite.merge(defaultFields),
  },
  query: {
    insert: insertQuery,
    get: z.discriminatedUnion('type', [totalCarbonEmbodimentQuery.merge(insertQueryShared).merge(defaultFields)]),
  },
  query_rpc: {
    insert: insertQueryRpc,
    get: insertQueryRpc.merge(defaultFields),
  },
  settings: {
    insert: insertSettings,
    get: insertSettings.extend({
      id: z.string(),
      created_at: z.date(),
      updated_at: z.date(),
    }),
  },
  organisation_registries: {
    insert: insertOrganisationRegistries,
    get: insertOrganisationRegistries.extend({
      id: z.string(),
    }),
  },
}

export type InsertConnection = z.infer<typeof Zod.connection.insert>
export type ConnectionRow = z.infer<typeof Zod.connection.get>
export type QueryRow = z.infer<typeof Zod.query.get>
export type SettingsRow = z.infer<typeof Zod.settings.get>
export type OrganisationRegistriesRow = z.infer<typeof Zod.organisation_registries.get>
export type TABLES_TUPLE = typeof tablesList
export type TABLE = TABLES_TUPLE[number]
export type Models = {
  [key in TABLE]: {
    get: z.infer<(typeof Zod)[key]['get']>
    insert: z.infer<(typeof Zod)[key]['insert']>
  }
}

export type ColumnsByType<M extends TABLE, T> = {
  [K in keyof Models[M]['get']]-?: Models[M]['get'][K] extends T ? K : never
}[keyof Models[M]['get']]

type WhereComparison<M extends TABLE> = {
  [key in keyof Models[M]['get']]: [
    Extract<key, string>,
    '=' | '>' | '>=' | '<' | '<=' | '<>' | 'LIKE' | 'ILIKE',
    Extract<Models[M]['get'][key], Knex.Value>,
  ]
}
export type WhereMatch<M extends TABLE> = {
  [key in keyof Models[M]['get']]?: Models[M]['get'][key]
}

export type Where<M extends TABLE> = WhereMatch<M> | (WhereMatch<M> | WhereComparison<M>[keyof Models[M]['get']])[]
export type Order<M extends TABLE> = [keyof Models[M]['get'], 'asc' | 'desc'][]
export type Update<M extends TABLE> = Partial<Models[M]['get']>

export type IDatabase = {
  [key in TABLE]: () => Knex.QueryBuilder
}

export default Zod
