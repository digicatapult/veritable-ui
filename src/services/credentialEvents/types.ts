import { Credential, CredentialFormatData } from '../../models/veritableCloudagent/internal.js'

export interface CredentialEventHandlerBase {
  handleProposalReceived(credential: Credential, formatData: CredentialFormatData): Promise<void>
  handleOfferReceived(credential: Credential, formatData: CredentialFormatData): Promise<void>
  handleRequestReceived(credential: Credential, formatData: CredentialFormatData): Promise<void>
  handleCredentialReceived(credential: Credential, formatData: CredentialFormatData): Promise<void>
  handleDone(credential: Credential, formatData: CredentialFormatData): Promise<void>
}

export interface CredentialEventHandler<EventName extends string, EventVersion extends string>
  extends CredentialEventHandlerBase {
  get schemaName(): EventName
  get schemaVersion(): EventVersion
}
