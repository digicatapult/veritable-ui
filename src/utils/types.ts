export type DiscriminateUnion<T, K extends keyof T, V extends T[K]> = T extends Record<K, V> ? T : never

export type MapDiscriminatedUnion<T extends Record<K, string>, K extends keyof T> = {
  [V in T[K]]: DiscriminateUnion<T, K, V>
}
