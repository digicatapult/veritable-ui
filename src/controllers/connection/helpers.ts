import { ILogger } from '../../logger.js'
import { ConnectionRow } from '../../models/db/types.js'

export const checkDb = async (rows: ConnectionRow[], initialPinAttemptsRemaining: number, logger: ILogger) => {
  const [connectionCheck] = rows
  if (initialPinAttemptsRemaining > 0 && connectionCheck.pin_tries_remaining_count == 0) {
    logger.debug('Maximum number of pin attempts has been reached.')
    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message:
        'Maximum number of pin attempts has been reached, please reach out to the company you are attempting to connect to.',
      nextScreen: 'error',
    }
  } else if (connectionCheck.status === 'verified_us') {
    logger.debug('Pin has been verified.')

    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message: 'Success',
      nextScreen: 'success',
    }
  } else if (
    connectionCheck.pin_tries_remaining_count < initialPinAttemptsRemaining ||
    (initialPinAttemptsRemaining === 0 && connectionCheck.pin_tries_remaining_count === 4)
  ) {
    logger.debug(`Pin was invalid remaining attempts number:${connectionCheck.pin_tries_remaining_count}.`)
    return {
      localPinAttempts: connectionCheck.pin_tries_remaining_count,
      message: `Sorry, your code is invalid. You have ${connectionCheck.pin_tries_remaining_count} attempts left before the PIN expires.`,
      nextScreen: 'form',
    }
  }
  logger.debug(`Pollig: No change in db detected.`)

  return false
}
