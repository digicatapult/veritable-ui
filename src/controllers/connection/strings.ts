export enum RegistryCountryCode {
  UK = 'UK',
  NY = 'NY',
}
