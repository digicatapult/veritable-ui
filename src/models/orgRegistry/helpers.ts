import z from 'zod'
import { InternalError } from '../../errors.js'
import { ILogger } from '../../logger.js'
import { RegistryType } from '../db/types.js'
import { OrganisationProfile, OrganisationRequest, Registries, SharedOrganisationInfo } from './organisationRegistry.js'
import { companyHouseResultSchema } from './registrySchemas/companyHouseSchema.js'
import { nyStateResultSchema } from './registrySchemas/NYStateSchema.js'
import { openCorporatesResultSchema } from './registrySchemas/openCorporatesSchema.js'

export const makeOrganisationRegistryRequest = async (
  route: string,
  registryName: string,
  apiKey: string | null
): Promise<unknown> => {
  const url = new URL(route)

  const headers: HeadersInit = {}
  if (apiKey !== null) {
    headers.Authorization = apiKey
  }

  const response = await fetch(url.toString(), {
    method: 'GET',
    headers,
  })
  const contentType = response.headers.get('content-type')
  if (response.ok) {
    if (contentType && contentType.includes('application/json')) {
      return response.json()
    }
    throw new InternalError(`Unexpected content-type ${contentType} from ${registryName} API, for ${url.toString()}`)
  }

  if (response.status === 404) {
    return null
  }

  throw new InternalError(`Error calling ${registryName} API`)
}
const retrieveEndpointAndSchema = (
  orgReq: OrganisationRequest,
  registry: Registries[RegistryType]
): { url: string; schema: z.ZodSchema; apiKey?: string } => {
  switch (orgReq.selectedRegistry) {
    case 'company_house':
      return {
        url: `${registry.url}/company/${encodeURIComponent(orgReq.companyNumber)}`,
        schema: companyHouseResultSchema,
        apiKey: registry.api_key,
      }

    case 'ny_state':
      return { url: `${registry.url}?dos_id=${orgReq.companyNumber}`, schema: nyStateResultSchema }

    case 'open_corporates':
      return {
        url: `${registry.url}/companies/${orgReq.registryCountryCode.toLowerCase()}/${encodeURIComponent(orgReq.companyNumber)}?api_token=${registry.api_key}`,
        schema: openCorporatesResultSchema,
      }

    default:
      throw new InternalError(`Registry ${orgReq.selectedRegistry} not set`)
  }
}

export const organisationRegistryProfile = async (
  orgReq: OrganisationRequest,
  registry: Registries[RegistryType],
  logger: ILogger
): Promise<OrganisationProfile> => {
  logger.info(`Getting ${registry.registry_name} profile for ${orgReq.companyNumber}`)
  const urlAndSchema = retrieveEndpointAndSchema(orgReq, registry)
  const companyProfile = await makeOrganisationRegistryRequest(
    urlAndSchema.url,
    registry.registry_name,
    urlAndSchema.apiKey ? urlAndSchema.apiKey : null
  )
  if (companyProfile === null) {
    return { type: 'notFound' }
  }
  try {
    const result = urlAndSchema.schema.parse(companyProfile)
    const sharedInfo = result as SharedOrganisationInfo
    return { type: 'found', company: sharedInfo }
  } catch (error) {
    logger.error(`Error parsing ${registry.registry_name} profile for ${orgReq.companyNumber}: ${error}`)
    return { type: 'notFound' }
  }
}
