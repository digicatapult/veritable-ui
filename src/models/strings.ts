import en from 'i18n-iso-countries/langs/en.json' with { type: 'json' }

/**
 * W3C Decentralized Identifier format v1.0
 * @pattern did:[A-Za-z0-9]+:[A-Za-z0-9]+
 * @example "did:key:z6Mkk7yqnGF3YwTrLpqrW6PGsKci7dNqh1CjnvMbzrMerSeL"
 */
export type DID = string

/**
 * @example "WgWxqztrNooG92RXvxSTWv:3:CL:20:tag"
 */
export type CredentialDefinitionId = string

/**
 * @example "WgWxqztrNooG92RXvxSTWv:2:schema_name:1.0"
 */
export type SchemaId = string

/**
 * @example "1.0.0"
 */
export type Version = string

/**
 * Stringified UUIDv4.
 * @pattern [0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89ABab][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}
 * @example "52907745-7672-470e-a803-a2f8feb52944"
 */
export type UUID = string

/**
 * Hex string with 0x prefix
 * @pattern 0x[0-9a-zA-Z]+
 * @format hex
 * @example "0xFF"
 */
export type HEX = `0x${string}`

/**
 * ISO 8601 date string
 * @pattern (\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))
 * @format date
 */
export type DATE = string

/**
 * Email format
 * @pattern (?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])
 * @format email
 * @example "user@example.com"
 */
export type EMAIL = string

/**
 * Pin Code format
 * @pattern ^[0-9]{6}$
 * @minLength 6
 * @maxLength 6
 * @example 123456
 */
export type PIN_CODE = string
export const pinCodeRegex = /^[0-9]{6}$/

/**
 * Company number format
 * @pattern ^(((AC|CE|CS|FC|FE|GE|GS|IC|LP|NC|NF|NI|NL|NO|NP|OC|OE|PC|R0|RC|SA|SC|SE|SF|SG|SI|SL|SO|SR|SZ|ZC|\d{2})\d{6})|((IP|SP|RS)[A-Z\d]{6})|(SL\d{5}[\dA]))$
 * @minLength 8
 * @maxLength 8
 * @example 07964699
 */
export type COMPANY_NUMBER = string
export const companyNumberRegex =
  /^(((AC|CE|CS|FC|FE|GE|GS|IC|LP|NC|NF|NI|NL|NO|NP|OC|OE|PC|R0|RC|SA|SC|SE|SF|SG|SI|SL|SO|SR|SZ|ZC|\d{2})\d{6})|((IP|SP|RS)[A-Z\d]{6})|(SL\d{5}[\dA]))$/

/**
 * Socrata company number format
 * @pattern ^\d{7}$
 * @example 3211809
 */
export type SOCRATA_NUMBER = string
export const socrataRegex = /^\d{7}$/

/**
 * Base64 url compatible string (see rfc4648 section-5)
 * @pattern ^[a-zA-Z0-9_\-]+$
 * @example VGhpcyBpcyBzb21lIGV4YW1wbGUgdGV4dA
 */
export type BASE_64_URL = string
export const base64UrlRegex = /^[a-zA-Z0-9_-]+$/

/**
 * Bank Identifier Code (BIC)
 * @pattern ^[A-Z]{4}[A-Z]{2}[A-Z0-9]{2}([A-Z0-9]{3})?$
 * @minLength 8
 * @maxLength 11
 * @example DEUTDEFF500
 */
export type BIC = string
export const bicRegex = /^[A-Z]{4}[A-Z]{2}[A-Z0-9]{2}([A-Z0-9]{3})?$/

export type CountryCode = keyof typeof en.countries
export const countryCodes = Object.keys(en.countries) as [CountryCode]
