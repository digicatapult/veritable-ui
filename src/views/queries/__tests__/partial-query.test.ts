import { expect } from 'chai'
import { describe } from 'mocha'
import { mockIds } from '../../../controllers/queries/__tests__/helpers.js'
import { ConnectionRow, QueryRow } from '../../../models/db/types.js'
import { CountryCode } from '../../../models/strings.js'
import QueryResponseTemplates from '../queryResponse.js'

const templates = new QueryResponseTemplates()
const sampleDate = new Date(Date.UTC(2024, 6, 4))
const connectionsExample: ConnectionRow[] = [{}, {}, {}, {}].map((_, i) => ({
  id: `${mockIds.connectionId.substr(0, mockIds.connectionId.length - 1)}${i}`,
  company_name: 'I own you',
  company_number: '3546783',
  status: 'verified_both',
  agent_connection_id: '233495875757',
  pin_attempt_count: 0,
  pin_tries_remaining_count: null,
  created_at: sampleDate,
  updated_at: sampleDate,
  registry_country_code: 'GB' as CountryCode,
}))

const queryExample: QueryRow = {
  id: 'aaaaaaaa-0001-0000-0000-d8ae0805059e',
  connection_id: 'cccccccc-0001-0000-0000-d8ae0805059e',
  type: 'total_carbon_embodiment',
  parent_id: '5390af91-c551-4d74-b394-d8ae0805059a',
  status: 'resolved',
  details: {
    subjectId: {
      idType: 'product_and_quantity',
      content: { productId: 'partial-product-id', quantity: 10 },
    },
  },
  response_id: null,
  response: null,
  role: 'requester',
  created_at: new Date(),
  updated_at: new Date(),
  expires_at: new Date(),
}

describe('Partial Query', () => {
  describe('if partial is set to false', () => {
    it('does not render connections list and leaves quantity input enabled', async () => {
      const rendered = await templates.queryResponsePage({
        formStage: 'form',
        connection: {
          company_name: 'VER123',
          company_number: '3456789',
          status: 'verified_both',
          id: mockIds.companyId,
          agent_connection_id: mockIds.agentConnectionId,
          pin_tries_remaining_count: null,
          pin_attempt_count: 0,
          created_at: new Date(),
          updated_at: new Date(),
          registry_country_code: 'GB' as CountryCode,
        },
        type: 'total_carbon_embodiment',
        query: queryExample,
        partial: false,
        connections: connectionsExample,
      })

      expect(rendered).to.matchSnapshot()
    })
  })

  it('renders partial query table with connections that are verified', async () => {
    const rendered = await templates.queryResponsePage({
      formStage: 'form',
      connection: {
        company_name: 'VER123',
        company_number: '3456789',
        status: 'verified_both',
        id: mockIds.companyId,
        agent_connection_id: mockIds.agentConnectionId,
        pin_tries_remaining_count: null,
        pin_attempt_count: 0,
        created_at: new Date(),
        updated_at: new Date(),
        registry_country_code: 'GB' as CountryCode,
      },
      type: 'total_carbon_embodiment',
      query: queryExample,
      partial: true,
      connections: connectionsExample,
    })
    expect(rendered).to.matchSnapshot()
  })
})
