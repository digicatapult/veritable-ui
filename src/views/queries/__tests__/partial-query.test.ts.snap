// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partial Query if partial is set to false does not render connections list and leaves quantity input enabled 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><script src=\\"/lib/htmx.org/htmx.min.js\\"></script><script src=\\"/lib/htmx-ext-json-enc/json-enc.js\\"></script><script src=\\"/public/scripts/auth-redirect.js\\"></script><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon-dark.ico\\" media=\\"(prefers-color-scheme: dark)\\"/><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon.ico\\" media=\\"(prefers-color-scheme: light), (prefers-color-scheme: no-preference)\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/main.css\\"/><title>Veritable - Select Company</title></head><body hx-ext=\\"json-enc\\"><nav id=\\"side-bar\\"><img src=\\"/public/images/logo-square.svg\\"/><div class=\\"profile-icon\\" style=\\"background-color:#38b6ff;\\">A</div><a title=\\"categories\\" href=\\"/\\" style=\\"--background-image:url('/public/images/category.svg');\\"></a><a title=\\"queries\\" href=\\"/queries\\" data-active style=\\"--background-image:url('/public/images/query.svg');\\"></a><a title=\\"credentials\\" href=\\"/credentials\\" style=\\"--background-image:url('/public/images/credential.svg');\\"></a><a title=\\"connections\\" href=\\"/connection\\" style=\\"--background-image:url('/public/images/connection.svg');\\"></a><a title=\\"settings\\" href=\\"/settings\\" style=\\"--background-image:url('/public/images/setting.svg');\\"></a></nav><main><div id=\\"content-header\\"><h1>Select Company To Send Your Query To</h1><div id=\\"content-header-nav\\"><a title=\\"home\\" class=\\"nav icon\\" href=\\"/\\"></a><a title=\\"Query Management\\" href=\\"/queries\\">Query Management</a><a title=\\"Query Request product-id-test\\" href=\\"/queries/scope-3-carbon-consumption/00000000-0000-0000-0000-d8ae0805059e/response\\">Query Request product-id-test</a></div></div><div id=\\"content-main\\"><div class=\\"connections header\\"></div><div class=\\"card-body\\"><div class=\\"container-scope3-carbon\\"><div class=\\"scope3-co2-left\\"><h1 id=\\"scope3-co2-heading\\">Scope 3 Carbon Consumption</h1><p style=\\"padding-right:50px;\\" class=\\"query-text-carbon3-consumption\\">Provide the total scope 3 carbon consumption for the specified products / component.</p><p class=\\"query-text-carbon3-consumption\\">If you do not have all the required information, please forward this query to your suppliers, to aggregate their responses before submitting the final total. </p></div><div class=\\"scope3-co2-right\\"><p class=\\"query-text-carbon3-consumption\\">What are the total Scope 3 carbon emissions for the product/component below?</p><div hx-swap-oob=\\"true\\" hx-swap=\\"ignoreTitle:true\\" id=\\"partial-query\\"><form id=\\"scope-3-carbon-consumption\\" hx-post=\\"/queries/scope-3-carbon-consumption/00000000-0000-0000-0000-d8ae0805059e/response\\" hx-select=\\"main > *\\" hx-include=\\"[name='quantity'], [name='companyId'], [name='productId']\\" hx-target=\\"main\\" hx-swap=\\"innerHTML\\"><p>Product ID: product-id-test<br/>Quantity: 2</p><input type=\\"hidden\\" name=\\"companyId\\" value=\\"cccccccc-0001-0000-0000-d8ae0805059e\\"/><div class=\\"input-container\\"><label for=\\"co2-emissions-input\\" class=\\"input-label\\">Total Scope 3 carbon emissions</label><input id=\\"co2-emissions-input\\" name=\\"emissions\\" placeholder=\\"Value in kg CO2e (to be aggregated)\\" class=\\"input-with-label\\" type=\\"text\\" required/></div><div class=\\"input-container\\"><input hx-trigger=\\"changed, click\\" hx-target=\\"#partial-query\\" hx-get=\\"/queries/00000000-0000-0000-0000-d8ae0805059e/partial\\" id=\\"partial-response-input\\" name=\\"partialQuery\\" type=\\"checkbox\\"/><label for=\\"partial-response-input\\">Partial response</label></div><p style=\\"font-style:italic;font-size:14px;;\\">*If partial response checkbox is ticked, you must share this query with another supplier in your network, where your responses will be aggregated.</p><br/><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"success\\"><span>Submit Response</span></button></form></div></div></div></div></div></main></body></html>"`;

exports[`Partial Query renders partial query table with connections that are verified 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><script src=\\"/lib/htmx.org/htmx.min.js\\"></script><script src=\\"/lib/htmx-ext-json-enc/json-enc.js\\"></script><script src=\\"/public/scripts/auth-redirect.js\\"></script><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon-dark.ico\\" media=\\"(prefers-color-scheme: dark)\\"/><link rel=\\"icon\\" type=\\"image/ico\\" href=\\"/public/images/favicon.ico\\" media=\\"(prefers-color-scheme: light), (prefers-color-scheme: no-preference)\\"/><link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"/public/styles/main.css\\"/><title>Veritable - Select Company</title></head><body hx-ext=\\"json-enc\\"><nav id=\\"side-bar\\"><img src=\\"/public/images/logo-square.svg\\"/><div class=\\"profile-icon\\" style=\\"background-color:#38b6ff;\\">A</div><a title=\\"categories\\" href=\\"/\\" style=\\"--background-image:url('/public/images/category.svg');\\"></a><a title=\\"queries\\" href=\\"/queries\\" data-active style=\\"--background-image:url('/public/images/query.svg');\\"></a><a title=\\"credentials\\" href=\\"/credentials\\" style=\\"--background-image:url('/public/images/credential.svg');\\"></a><a title=\\"connections\\" href=\\"/connection\\" style=\\"--background-image:url('/public/images/connection.svg');\\"></a><a title=\\"settings\\" href=\\"/settings\\" style=\\"--background-image:url('/public/images/setting.svg');\\"></a></nav><main><div id=\\"content-header\\"><h1>Select Company To Send Your Query To</h1><div id=\\"content-header-nav\\"><a title=\\"home\\" class=\\"nav icon\\" href=\\"/\\"></a><a title=\\"Query Management\\" href=\\"/queries\\">Query Management</a><a title=\\"Query Request product-id-test\\" href=\\"/queries/scope-3-carbon-consumption/query-id-test/response\\">Query Request product-id-test</a></div></div><div id=\\"content-main\\"><div class=\\"connections header\\"></div><div class=\\"card-body\\"><div class=\\"container-scope3-carbon\\"><div class=\\"scope3-co2-left\\"><h1 id=\\"scope3-co2-heading\\">Scope 3 Carbon Consumption</h1><p style=\\"padding-right:50px;\\" class=\\"query-text-carbon3-consumption\\">Provide the total scope 3 carbon consumption for the specified products / component.</p><p class=\\"query-text-carbon3-consumption\\">If you do not have all the required information, please forward this query to your suppliers, to aggregate their responses before submitting the final total. </p></div><div class=\\"scope3-co2-right\\"><p class=\\"query-text-carbon3-consumption\\">What are the total Scope 3 carbon emissions for the product/component below?</p><div hx-swap-oob=\\"true\\" hx-swap=\\"ignoreTitle:true\\" id=\\"partial-query\\"><form id=\\"scope-3-carbon-consumption\\" hx-post=\\"/queries/scope-3-carbon-consumption/query-id-test/response\\" hx-select=\\"main > *\\" hx-include=\\"[name='quantity'], [name='companyId'], [name='productId']\\" hx-target=\\"main\\" hx-swap=\\"innerHTML\\"><p>Product ID: product-id-test<br/>Quantity: 2</p><input type=\\"hidden\\" name=\\"companyId\\" value=\\"cccccccc-0001-0000-0000-d8ae0805059e\\"/><div class=\\"input-container\\"><label for=\\"co2-emissions-input\\" class=\\"input-label\\">Total Scope 3 carbon emissions</label><input id=\\"co2-emissions-input\\" name=\\"emissions\\" placeholder=\\"Value in kg CO2e (to be aggregated)\\" class=\\"input-with-label\\" type=\\"text\\" required/></div><div class=\\"input-container\\"><input hx-trigger=\\"changed, click\\" hx-target=\\"#partial-query\\" hx-get=\\"/queries/query-id-test/partial\\" id=\\"partial-response-input\\" name=\\"partialQuery\\" type=\\"checkbox\\" checked/><label for=\\"partial-response-input\\">Partial response</label></div><p style=\\"font-style:italic;font-size:14px;;\\">*If partial response checkbox is ticked, you must share this query with another supplier in your network, where your responses will be aggregated.</p><div class=\\"query-partial-container list-page\\"><table><thead><th>Select</th><th>Company Name</th><th>Product ID</th><th>Quantity</th></thead><tbody hx-swap-oob=\\"true\\"><tr id=\\"tr-cccccccc-0000-0000-0000-d8ae08050590\\" hx-swap-oob=\\"true\\"><input name=\\"connectionIds\\" type=\\"hidden\\" value=\\"cccccccc-0000-0000-0000-d8ae08050590\\" disabled/><td><input name=\\"partialSelect\\" type=\\"checkbox\\" hx-trigger=\\"click\\" hx-get=\\"/queries/partial-select/cccccccc-0000-0000-0000-d8ae08050590\\" hx-target=\\"#tr-cccccccc-0000-0000-0000-d8ae08050590\\"/></td><td>I own you</td><td><input name=\\"productIds\\" placeholder=\\"Product ID\\" class=\\"input-basic disabled\\" type=\\"text\\" disabled/></td><td><input name=\\"quantities\\" placeholder=\\"Quantity\\" class=\\"input-basic disabled\\" type=\\"number\\" disabled/></td></tr><tr id=\\"tr-cccccccc-0000-0000-0000-d8ae08050591\\" hx-swap-oob=\\"true\\"><input name=\\"connectionIds\\" type=\\"hidden\\" value=\\"cccccccc-0000-0000-0000-d8ae08050591\\" disabled/><td><input name=\\"partialSelect\\" type=\\"checkbox\\" hx-trigger=\\"click\\" hx-get=\\"/queries/partial-select/cccccccc-0000-0000-0000-d8ae08050591\\" hx-target=\\"#tr-cccccccc-0000-0000-0000-d8ae08050591\\"/></td><td>I own you</td><td><input name=\\"productIds\\" placeholder=\\"Product ID\\" class=\\"input-basic disabled\\" type=\\"text\\" disabled/></td><td><input name=\\"quantities\\" placeholder=\\"Quantity\\" class=\\"input-basic disabled\\" type=\\"number\\" disabled/></td></tr><tr id=\\"tr-cccccccc-0000-0000-0000-d8ae08050592\\" hx-swap-oob=\\"true\\"><input name=\\"connectionIds\\" type=\\"hidden\\" value=\\"cccccccc-0000-0000-0000-d8ae08050592\\" disabled/><td><input name=\\"partialSelect\\" type=\\"checkbox\\" hx-trigger=\\"click\\" hx-get=\\"/queries/partial-select/cccccccc-0000-0000-0000-d8ae08050592\\" hx-target=\\"#tr-cccccccc-0000-0000-0000-d8ae08050592\\"/></td><td>I own you</td><td><input name=\\"productIds\\" placeholder=\\"Product ID\\" class=\\"input-basic disabled\\" type=\\"text\\" disabled/></td><td><input name=\\"quantities\\" placeholder=\\"Quantity\\" class=\\"input-basic disabled\\" type=\\"number\\" disabled/></td></tr><tr id=\\"tr-cccccccc-0000-0000-0000-d8ae08050593\\" hx-swap-oob=\\"true\\"><input name=\\"connectionIds\\" type=\\"hidden\\" value=\\"cccccccc-0000-0000-0000-d8ae08050593\\" disabled/><td><input name=\\"partialSelect\\" type=\\"checkbox\\" hx-trigger=\\"click\\" hx-get=\\"/queries/partial-select/cccccccc-0000-0000-0000-d8ae08050593\\" hx-target=\\"#tr-cccccccc-0000-0000-0000-d8ae08050593\\"/></td><td>I own you</td><td><input name=\\"productIds\\" placeholder=\\"Product ID\\" class=\\"input-basic disabled\\" type=\\"text\\" disabled/></td><td><input name=\\"quantities\\" placeholder=\\"Quantity\\" class=\\"input-basic disabled\\" type=\\"number\\" disabled/></td></tr></tbody></table></div><br/><button class=\\"button  \\" data-variant=\\"filled\\" type=\\"submit\\" name=\\"action\\" value=\\"success\\"><span>Submit Response</span></button></form></div></div></div></div></div></main></body></html>"`;
